### ISSUE TITLE
Feature request: make it possible to keep docker container warm

### ISSUE THREAD

**Issue by jandockx** [aws/aws-sam-cli#239]
I understand from other issues that a new docker container is started for each request. This makes some experiments or automated tests undoable in practice. SAM Local is much too slow in the context where more then 1 request is to be handled.

I suspect that hot reloading depends on this feature.

I think it would be a good idea to make it possible to choose, while this project evolves further, to forego hot reloading, but to keep the docker container warm.

Something like 

    sam local start-api -p <PORT> --profile <AWS PROFILE> --keep-it-warm

This would broaden the applicability of sam local enormously.

Thank you for considering this suggestion. This looks like an awesome project.

**Comment by aldegoeij** [comment]
+1 Python container takes too long to start for simple debugging...

**Comment by zippadd** [comment]
+1. This currently makes local automated testing painful at best.

Thanks for the continued work on this project!

**Comment by dannymcpherson** [comment]
Have there been any eyes on this?  The benefit would be so huge.

**Comment by cagoi** [comment]
+1

**Comment by daveykane** [comment]
+1

**Comment by adrian1358** [comment]
+1

**Comment by CRogers** [comment]
+1, even a simple hello world java8 lambda takes 3/4 seconds for each request!

**Comment by monofonik** [comment]
+1

**Comment by luisvsm** [comment]
+1 Very interested in this feature

**Comment by luketn** [comment]
+1 Yes please!

**Comment by nodeit** [comment]
+1, throwing my hat in the ring on this too

**Comment by jfuss** [comment]
As a note: Please use the reaction feature on the top comment. We do look at issues sorted by thumbs up (as well as other reactions). Commenting +1 does not good for that and adds noise to the issue.

**Comment by scoates** [comment]
@jfuss I agree (and had done this). Any feedback from your team would be helpful here, though. The closest thing we had to knowing if this is on your radar (before your comment) was duplicate issue consolidation and labeling.

**Comment by ejoncas** [comment]
+1, this would be very beneficial for people using java + spring boot.

**Comment by thoratou** [comment]
+1, around 1s for golang case

**Comment by peachepe** [comment]
1 second is a world's difference when building an API and you expect to serve more than 1 request.

I think it's well worth the feature.

**Comment by sanathkr** [comment]
@kevanpng Hey I was looking through your code to understand what exactly you did.. So basically, you create the container once with a fixed name, run the function, and on next invocation look for container with same name and simply `container.exec_run` instead of creating it from scratch again. Is my summary correct?

I am super surprised Docker container creation makes this big of a difference. We can certainly look deeper into this if it is becoming usability blocker.

**Comment by scoates** [comment]
@sanathkr. Thanks for looking at this. FWIW, it's a huge usability blocker for me:

```
~/src/faculty/buildshot$ time curl -s http://127.0.0.1:3000/ >/dev/null # SAM container via Docker

real	0m6.891s
user	0m0.012s
sys	0m0.021s
~/src/faculty/buildshot$ time curl -s http://127.0.0.1:5000/ >/dev/null # regular python app via flask dev/debug server (slow)

real	0m0.039s
user	0m0.012s
sys	0m0.019s
```

And the `Instancing..` is quick. It's Docker (and the way Docker is used here) that's slow. The (slow) werkzeug-based dev server is ~175x faster than waiting around for Docker. And this is for every request, not just startup.

**Comment by sanathkr** [comment]
I did some more profiling by crudely commenting out parts of the codebase. Also this is not run multiple times. So the numbers are ballpark estimates. I ran `sam init` and ran `sam local-start-api` on a simple HelloWorld Lambda function created by the init template.

Platform: MacOSX
Docker version: 18.06.0

Total execution time (`sam local start-api`): 2.67 seconds
Skip pull images (`sam local start-api --skip-pull-image`): 1.45 seconds
Create container, run it, and return immediately without waiting for function terminate: 1.05 seconds 
Create container, don't run it: 0.2 seconds
SAM CLI code overhead (don't create container at all): 0.045 seconds

Based on the above numbers, I arrived at a rough estimate for each step of the invoke path by assuming:
 
Total execution = SAM CLI overhead + Docker Image pull + Create container + Run Container + Run function

Then, here is how much each steps took:

**SAM CLI Overhead**: 0.045 seconds
**Docker Image Pull Check**: 1.3 seconds
**Create Container**: 0.15 seconds
**Run container**: 0.85 seconds
**Run function**: 0.45 seconds

The most interesting part is Create vs Run container durations. Run is 5x of Create. So it is better if we optimized for the Run duration.

If we were to do a warm start, then we would be saving some fraction of the **0.85 seconds** it took to run the container. We should be keeping the runtime process up and running inside the container and re-run just the function in-place. Otherwise we aren't going to save much.

**Comment by vrealzhou** [comment]
+1

**Comment by sanathkr** [comment]
@scoates  Thank you for the detailed information. Definitely worth instrumenting the code further to dive deep. I will reach out when I have an instrumented version of SAM CLI to try out.

**Comment by gkfirst8** [comment]
In case of java8: the on-line version of Lambda keeps the JVM running for a long time. It would be nice to replicate this in 'sam local' mode as then I can study any memory usage issues/OutOfMemoryExceptions I see.

**Comment by clicktravel-steffan** [comment]
+1 We're running test suites for fairly simple microservices where individual tests can take 10+ seconds to run.

**Comment by vrealzhou** [comment]
I've wrote a tool can simulate sam local start-lambda and keep warm for Lambda written in Golang.
https://github.com/vrealzhou/lambda-local. Not perfect but helps in our own project.

**Comment by metaskills** [comment]
> Maybe it could be that docker is slow on Mac?

I've done some reading around on the internet and think this is just the case. For example, when requiring a bunch of gems, things slow down locally ALOT. And it appears this slowness comes from mounted volumes and Docker's `osxfs`. No clue how to make this better tho.

**Comment by metaskills** [comment]
I've been reading this article (https://docs.docker.com/docker-for-mac/osxfs-caching/) with regard to Docker volume performance on Mac. I've been reading a little of SAM's Python (not my primary language) and I've landed in `local/docker/manager.py` and wonder if we should be using `:delegated` or `:cached` along with `--docker-volume-basedir, -v, SAM_DOCKER_VOLUME_BASEDIR` options to help speed this up?

**Comment by metaskills** [comment]
OK, not a Python person, but I was able to hack up a few things in my local `container.py` source and change this from `ro` to `delegated`.

```python
        LOG.info("Mounting %s as %s:delegated inside runtime container", self._host_dir, self._working_dir)

        kwargs = {
            "command": self._cmd,
            "working_dir": self._working_dir,
            "volumes": {
                self._host_dir: {
                    # Mount the host directory as "read only" directory inside container at working_dir
                    # https://docs.docker.com/storage/bind-mounts
                    # Mount the host directory as "read only" inside container
                    "bind": self._working_dir,
                    "mode": "delegated"
                }
            },
            # We are not running an interactive shell here.
            "tty": False
        }
```

And this line in my `app.rb` went from `~45s` to `~13s`. Still not ideal, but better!

```ruby
puts(Benchmark.realtime { require_relative 'config/application' })
```

**Comment by metaskills** [comment]
I created this pull request that adds `delegated` after `ro` to the volume. https://github.com/awslabs/aws-sam-cli/pull/1046

**Comment by emorneau** [comment]
+1, my issue is not performance but the fact that we can't perform unit/acceptance tests on warm local lambda and bugs can be found after deployment.

**Comment by dsanders11** [comment]
@sanathkr, any progress here? It certainly makes local development a bit painful with how long the response times are.

I played around with some benchmarking myself and figured I'd share. I'm running Docker in a Linux VM. All timings include using `delegated` for the volume mount.

Just running a simple 'hello world' container which automatically removes once it's done takes ~300-400ms, which is on par with [what this source says as well](https://blog.iron.io/the-overhead-of-docker-run/). I used `docker events` to get accurate timestamps. Unfortunately that's not super fine-grained, going from `network connect` to `container start`. Considering I also tried with network set to none and there was no noticeable difference in the time between those two events, I'm sure more is happening between the first and the second, it's just not given a name.

I tried taking the volume mount of the picture and saw significant speed-up in execution duration from doing so. I simply made a Docker image based on `lambci/lambda:nodejs8.10` and added the directory to `/var/tmp`, then changed `container.py` to not do the volume mount, and hard-coded it to that image.

I saw a 3x speedup in execution duration when using this method (1500ms -> 500ms, 2200ms -> 700ms, etc). Note that this is from SAM's outputted duration, the real-world time includes the Docker overhead of ~400ms. So, the volume mount is definitely a source of significant slowdown in the container. A warm container would perform better if it copied the data and used the copied version instead of using directly from the mount. Of course, this would pose some problems for the hot-reload functionality, as it would require polling to ensure the copied version is up-to-date.

If it's preferred to keep each request in a single container that executes then exits, perhaps a pool of 'warm' containers could be kept. Say a pool of 5 (configurable by a flag) containers are created and each is only 'ready' once it is past the Docker overhead of creating the container *and* copying the data from the volume mount to `/var/task`. As the containers are used, new ones are brought up, keeping the pool at the correct size. Ideal pool size would be large enough to cover concurrent requests. The only real tricky thing here is containers would need to be killed off and recreated as file changes happen, otherwise hot-reload would be broken.

I think that would lead to the best performance since container creation overhead, and volume mount overhead seem significant. Volume mount overhead alone is probably enough to keep response times over 1 second most of the time.

**Comment by hoang-innomizetech** [comment]
Does anyone have any workaround to speed up the execution time?

**Comment by humank** [comment]
+1, considering when you run lambda function call, and invoke each other in local, there will be several entire lambda runtime creation life-cycle there, and it's takes long time.

**Comment by vtirta** [comment]
+1

**Comment by jfuss** [comment]
Hey all, 

I just opened a PR that should really help the pain of 'slow'/'warm' invokes. I would love for people to chime in on the PR (I have deeper details of caveats and approach in the description of #1305). For any willing parties that want to give it a try and install the pr from source, I would love your feedback. We feel it is important for this change to be tested with real workflows to make sure we are hitting the need of you all. Thanks! 🎉

**Comment by frankh** [comment]
I've also made an attempt at this (#1319) - differs in that it keeps the container running constantly and invokes using `docker exec` instead of starting/stopping the container constantly.

lowers api gateway lambda latency from ~3s to ~0.2s

**Comment by thecaddy** [comment]
+1

**Comment by kastork** [comment]
The speed/time discussions here are all right on point - micronaut has largely solved the startup time issues that traditionally have made lambdas with JVM functions somewhat impractical, but the fresh-container-per-invocation is still a large impact.

But there's another consideration that I've recently run into, and that is the case where your lambda caches data for use between warm-starts.

For example, my app makes AWS SDK calls to secrets manager and other things that I want to avoid when possible.  So on a cold start I make those calls, but on a warm start I don't.  The current situation makes testing such things impossible in local.

**Comment by hoang-innomizetech** [comment]
+100

**Comment by ivan589** [comment]
+1

**Comment by fondberg** [comment]
+1

**Comment by mhart** [comment]
FWIW, I'm going to be pushing out some changes to https://github.com/lambci/docker-lambda in the next week that should support this out of the box.

Basically, each container can be configured (via an env var) to expose an HTTP interface, specifically the [Lambda API](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) – which means (after the first invoke), each invoke can hit a warm Lambda. Current testing on my Mac gives around 433 req/sec (2.3ms/req).

Assuming you have an `index.js` with a `handler()` in your current directory, you can try it out with:

```
docker run --rm -v $PWD:/var/task -e STAY_OPEN=1 -p 3000:3000 lambci/lambda:nodejs8.10-beta index.handler
```

(it won't print anything, but if it stays running... it's working)

Then you can invoke it (multiple times) using:

```
aws lambda invoke --endpoint http://localhost:3000 --no-sign-request --function-name doesnotmatter --payload '{}' output.json
```

Or simply just:

```
curl -d '{}' http://localhost:3000/2015-03-31/functions/myfunction/invocations
```

Very likely that the env var name will change, and probably the default port too – so don't get attached to those.

Will update you when I've pushed everything up 👍

**Comment by dgergely** [comment]
@mhart 
Hi! 
any update on this?
Thanks!

**Comment by mhart** [comment]
@dgergely yup, plenty – 1.2k lines of Go, C#, Java, Python and JS to be precise: https://github.com/lambci/docker-lambda/pull/218

**Comment by mhart** [comment]
Already, 1.8k lines later, all runtimes are supported. I still want to clean up some code and make sure stdout/stderr are supported correctly in the new model.

I've pushed all images to `-beta` tags. Eg, `lambci/lambda:ruby2.5-beta`, `lambci/lambda:dotnetcore2.1-beta`, etc

The env var to use is `DOCKER_LAMBDA_STAY_OPEN` and the port by default is `9001` – going to make this configurable soon.

So basically:

```
docker run --rm \
  -e DOCKER_LAMBDA_STAY_OPEN=1 \
  -p 9001:9001 \
  -v $PWD:/var/task \
  lambci/lambda:ruby2.5-beta \
  lambda_function.lambda_handler
```

You should then see:

```
Lambda API listening on port 9001...
```

And then you can invoke using:

```
curl -d '{}' http://localhost:9001/2015-03-31/functions/myfunction/invocations
```

OR

```
aws lambda invoke --endpoint http://localhost:9001 --no-sign-request --function-name myfunction --payload '{}' output.json
```

If you don't supply `DOCKER_LAMBDA_STAY_OPEN` then everything should function as it currently does (some very minor changes for some runtimes, but none of them should be breaking)

**Comment by mhart** [comment]
Due to the way I've implemented it, sharing the api server among all the runtimes, it's slower than the custom implementation in my earlier test. I now get around 70 req/s. But this is still much faster than cold starting each time. It would be possible to optimize this, but potentially at the cost of maintainability.

**Comment by mhart** [comment]
Support for warm invokes has been pushed to all docker-lambda runtimes 🎉

The documentation above still stands, ie invoke with:

```
docker run --rm \
  -e DOCKER_LAMBDA_STAY_OPEN=1 \
  -p 9001:9001 \
  -v $PWD:/var/task \
  lambci/lambda:ruby2.5 \
  lambda_function.lambda_handler
```

All runtimes also have support for `X-Amz-Log-Type: Tail` (`--log-type Tail` if invoking from the `aws` CLI), as well as `X-Amz-Invocation-Type: DryRun` (`--invocation-type DryRun`) and `X-Amz-Invocation-Type: Event` (`--invocation-type Event`).

All old images are available at `lambci/lambda:20191117-<runtime>`, eg `lambci/lambda:20191117-dotnetcore2.1` – in case ppl are encountering issues with the new images. However, I tried hard to ensure that sam invoke local should still function largely as it does today.

**Comment by mhart** [comment]
So, all that's left now is support for these warm invokes in `aws-sam-cli` 😸

**Comment by ranjan-purbey** [comment]
Our team is just getting started with lambda but one of the initial roadblocks we encountered was the slow response from local API Gateway instance created using aws-sam-cli. On a system with 16GB memory, each invocation takes ~7 seconds. This makes development really painful.
Any estimates on how long before the feature is integrated into SAM CLI?

**Comment by ranjan-purbey** [comment]
@mhart  In the command:
```
curl -d '{}' http://localhost:9001/2015-03-31/functions/myfunction/invocations
```
What does the URL depend on? I get 404 response from the server.

**Comment by mhart** [comment]
@ranjan-purbey use something like https://facebook.github.io/watchman/ – just restart the process whenever one of your files change

**Comment by mhart** [comment]
Added more documentation here: https://github.com/lambci/docker-lambda#running-in-stay-open-api-mode

**Comment by mhart** [comment]
@ranjan-purbey I added some documentation for developing and restarting whenever there are changes to your code: https://github.com/lambci/docker-lambda/#developing-in-stay-open-mode

**Comment by mhart** [comment]
I've actually added a watch mode to docker-lambda itself, instead of needing to rely on external file watchers to do the job for you. Just need to pass in `DOCKER_LAMBDA_WATCH=1` to activate.

Update documentation here: https://github.com/lambci/docker-lambda#developing-in-stay-open-mode

You can also manually reload the handler by passing `SIGHUP` to the container.

**Comment by langboost** [comment]
While waiting on the fix, it's helpful to know that the docker pull command (as @sanathkr  mentioned above) can be skipped, and accounts for about half of the waiting.

My personal experience with local api gateway testing is that I can shave about 5 seconds off per request by simply passing ```--skip-pull-image``` on launch.

```sam local start-api --skip-pull-image```

That's a very simple fix you can make to your dev workflow to save some pain for now. Thanks @sanathkr !

**Comment by literakl** [comment]
> @ranjan-purbey I added some documentation for developing and restarting whenever there are changes to your code: https://github.com/lambci/docker-lambda/#developing-in-stay-open-mode

Great work. I do not understand if this is intended for single API method, or complete API? Do I have to start Docker for every API method?

**Comment by OFranke** [comment]
If `sam` is using the same docker image under the hood, would it theoretically be possible to just pass the `DOCKER_LAMBDA_STAY_OPEN=1` variable via sam `environments.json`? 
Right now I observed that for some reason I cannot randomly add any variables to `environments.json`, just some that I defined before in the template.yaml.

**Comment by flache** [comment]
Are there any updates or is there a timeline on this? This is the single biggest blocker for us (and I can imagine for many others) to do more with AWS Lambda because this makes it almost impossible to develop and test stuff locally. Even with `--skip-pull-image`, a delay of ~5 seconds for each request makes it just unusable. Also there is the problem with global context not being preserved.

I understand that features must be prioritized but I am having a hard time to understand that everything that is running on lambda cannot be tested locally is not a high priority issue. Or am I missing something?

**Comment by literakl** [comment]
I have solved this trouble by moving away from Lambda to Node Express

Dne st 15. 4. 2020 12:50 uživatel flache <notifications@github.com> napsal:

**Comment by jfuss** [comment]
Update: The team is working on other prioritizes at the moment. We know the time it takes for invoking locally is a pain point for many and we have plans to address it in the future. We do not have an ETA as of now.

**Comment by OFranke** [comment]
@flache 
I've moved away from sam as it seems to not play so well with cdk at the moment, see https://github.com/awslabs/aws-sam-cli/issues/1911. I worked around it having an app that I run on docker locally but let cdk deploy it. Therefore  I just use different application entries, which are not so different at all.
```
// lambda entry
import { Response, Request } from 'express';

const awsServerlessExpress = require('aws-serverless-express');
const express = require('express');

const app = express();
const handler = (req: Request, res: Response): void => {
  try {
    app(
      req,
      res,
      (err: { status: number; statusCode: number; message: string }) => {
        if (err) {
          if (!res.headersSent) {
            res.statusCode = err.status || err.statusCode || 500;
            res.setHeader('Content-Type', 'application/json');
          }
          res.end(JSON.stringify({ errors: [{ message: `${err.message}` }] }));
          return;
        }
        if (!res.finished) {
          if (!res.headersSent) {
            res.statusCode = 404;
          }
          res.end(`'${req.url}' not found`);
        }
      },
    );
  } catch (err) {
    res.end(JSON.stringify({ errors: [{ message: `${err.message}` }] }));
  }
};

const server = awsServerlessExpress.createServer(handler, undefined);
exports.handler = (event: unknown, context: unknown): unknown =>
  awsServerlessExpress.proxy(server, event, context);
```

```
// docker entry
import express from 'express';

const main = async () => {
  const app = express();

  app.listen(5000, '0.0.0.0');
};

try {
  void main();
} catch (e) {
  console.error('Fatal error occurred starting server!');
  console.error(e);
  process.exit(101);
}
```

I have built a whole graphql service like that, and run it for a few weeks on AWS now. Seems to be fine.

**Comment by elthrasher** [comment]
For those who are very comfortable with Docker and docker-compose, I created a proxy image that works with the underlying SAM (lambci) images and can bring your lambda function into existing docker-compose workflows as a long-lived function. https://github.com/elthrasher/http-lambda-invoker

**Comment by literakl** [comment]
I have personally switched from AWS Lamda to NodeJS+Express+nodemon and my productivity and happiness boosted.

**Comment by duartemendes** [comment]
Spent the last week writing a CLI tool to help with this issue, just 2 days ago I published the first version.

It's available in [npm](https://www.npmjs.com/package/aws-sam-api-proxy) for download and installation. It provides both `DOCKER_LAMBDA_STAY_OPEN` and `DOCKER_LAMBDA_WATCH` environment variables to the underlying containers, mitigating cold starts after the first invocation and watching code changes.
 
I think the tool is easy to use (takes one command to run your api locally) but it's in a very early stage. It works very well for my APIs but I'm pretty sure I didn't take all use cases into consideration. So, give it a go, report any issues you find and please leave some feedback.

**Comment by S-Cardenas** [comment]
@duartemendes that tool is amazing! Congratulations and let me know if you need any help.

Does your tool currently support layers?

**Comment by duartemendes** [comment]
Thanks @S-Cardenas. It doesn't but it's something I'm happy to take a loot at 👍

**Comment by kingferiol** [comment]
This is really a road blocker for this technology for us. Too painfully.

It is not sustainable to wait 10 seconds per each request during development. Without any action on this, I think that we have to reconsider our approach to this technology.

**Comment by jfuss** [comment]
Update: We have prioritized some work that will help with the slow request time and provided a better warm invoke experience. I do not have timelines or ETAs to share at this point but wanted to communicate that we are starting to look at what we can do in this space.

**Comment by iballard95** [comment]
@jfuss any updates?

**Comment by guichafy** [comment]
I'm very excited to see this feature.

**Comment by leonardobork** [comment]
@jfuss any news?

**Comment by S-Cardenas** [comment]
Ditto. Would be great if this was officially released. Currently using [https://github.com/elthrasher/http-lambda-invoker](https://github.com/elthrasher/http-lambda-invoker) as a substitute.

**Comment by OGoodness** [comment]
🤞 Let's hope we can see this soon

**Comment by S-Cardenas** [comment]
Seems like it's getting very close to being approved and merged. Would love to get a notification when/if it does.

**Comment by millsy** [comment]
Fingers crossed this is soon added

**Comment by kaarejoergensen** [comment]
This feature has been added to the newest release (https://github.com/aws/aws-sam-cli/releases/tag/v1.14.0) 🎉

**Comment by mndeveci** [comment]
(As @kaarejoergensen mentioned 😄 ) Happy to inform that, this has been released with [v1.14](https://github.com/aws/aws-sam-cli/releases/tag/v1.14.0), resolving the issue.
